hail/src/main/scala/is/hail/expr/ir/agg/StagedBlockLinkedList.scala

```diff
@@ -0,0 +1,242 @@
+      elt.setup,
+      elt.m.mux(
+        pushMissing(lastNode),
+        pushPresent(lastNode, { dst =>
+          if(elemType.isPrimitive)
+            Region.storePrimitive(elemType, dst)(elt.value)
+          else
+            StagedRegionValueBuilder.deepCopy(fb, r, elemType, elt.value, dst)
+        })),
+      totalCount := totalCount + 1)
+  }
+
+  def append(r: Code[Region], bll: StagedBlockLinkedList): Code[Unit] = {
+    assert(bll.elemType.isOfType(elemType))
+    bll.foreach(push(r, _))
+  }
+
+  def writeToSRVB(srvb: StagedRegionValueBuilder): Code[Unit] = {
+    assert(srvb.typ.isOfType(bufferType))
+    Code(
+      srvb.start(totalCount, init = true),
+      foreach { elt =>
+        Code(
+          elt.m.mux(
+            srvb.setMissing(),
+            srvb.addWithDeepCopy(elemType, elt.value)),
+          srvb.advance())
+      })
+  }
+
+  def serialize(r: Code[Region], ob: Code[OutputBuffer]): Code[Unit] = {
+    Code(
+      foreachNode { n =>
+        // NOTE: we are only copying a portion of the missingness bytes, only as many as there are
+        // present (specifically: 'count(n)', not 'capacity(n)'). we can't use emitArray because it
+        // would mistakenly copy elements according to the capacity ...
+        val nMissingBytes = (count(n) + 7) >> 3
+        val writeMissingBytes = ob.writeBytes(r, buffer(n) + 4, nMissingBytes)
+        val eltm = bufferType.isElementMissing(r, buffer(n), i)
+        val eltOff = bufferType.elementOffsetInRegion(r, buffer(n), i)
+        val writeElt = EmitPackEncoder.emit(elemType, elemType, fb.apply_method, r, eltOff, ob)
+        Code(
+          ob.writeBoolean(true),
+          ob.writeInt(count(n)),
+          if(elemType.required) Code._empty else writeMissingBytes,
+          i := 0,
+          Code.whileLoop(i < count(n), Code(
+            if(elemType.required) writeElt else eltm.mux(Code._empty, writeElt),
+            i := i + 1)))
+      },
+      ob.writeBoolean(false))
+  }
+
+  def deserialize(r: Code[Region], ib: Code[InputBuffer]): Code[Unit] = {
+    // ... however, we can use EmitPackDecoder to decode an entire array and add it as a node
+    val srvb = new StagedRegionValueBuilder(EmitRegion(fb.apply_method, r), bufferType)
+    val bufFType = bufferType.fundamentalType
+    Code.whileLoop(ib.readBoolean(),
+      EmitPackDecoder.emitArray(bufFType, bufFType, fb.apply_method, ib, srvb),
+      appendShallow(r, srvb.end()))
+  }
+
+  private def appendShallow(r: Code[Region], aoff: Code[Long]): Code[Unit] = {
+    val len = bufferType.loadLength(r, aoff)
+    Code(
+      tmpNode := r.allocate(nodeType.alignment, nodeType.byteSize),
+      initNode(tmpNode,
+        buf = aoff,
+        count = len),
+      setNext(lastNode, tmpNode),
+      lastNode := tmpNode,
+      totalCount := totalCount + len)
+  }
+
+  def initWithDeepCopy(r: Code[Region], other: StagedBlockLinkedList): Code[Unit] = {
```

Comment:
Would it be possible to write this as a `deepCopy` of each buffer of `other`? I'd expect that to be significantly faster.